### ISSUE TITLE
error: [dereference.of.nullable] on an access from an annonymous class of an already initialized final field

### ISSUE THREAD

**Issue by mihai-dumitresq** [typetools/checker-framework#903]
I'm getting `error: [dereference.of.nullable] dereference of possibly-null reference` on an access from an annonymous class of an already initialized final field.
I'm using version 2.0.0.

```Java
import com.google.common.util.concurrent.SettableFuture;
class Foo {
  private final SettableFuture<Bar> mBar;
  public Foo() {
    mBar = SettableFuture.create();
    mBar.addListener(
        new Runnable() {
          @Override
          public void run() {
            // error: [dereference.of.nullable] dereference of possibly-null reference
            if (mBar.isCancelled()) {
              // whatever
            }
          }
        },
        MoreExecutors.directExecutor());
  }
}
```
Is there a way to annotate it (other than `@SuppressWarnings("dereference.of.nullable")` to make it work?

I've tried
- annotating `run()` with `@RequiresNonNull("mBar")` or `@RequiresNonNull("Foo.this.mBar")` but that fails:
```
error: [contracts.precondition.override.invalid] @NonNull void run(@Initialized @NonNull <anonymous Foo$1> this) in <anonymous Foo$1> cannot override @NonNull void run(@Initialized @NonNull Runnable this) in java.lang.Runnable; the precondition '@NonNull' for 'this.mBar' in the subclass method must be weaker (or equally strong) compared to the superclass method
          public void run() {
                      ^
```
- `run(/*@UnknownInitialization(Foo.class) Foo Foo.this*/)` (I'm probably doing this wrong, is there a right way?) but that results in:
```
error: the receiver type does not match the enclosing class type
              /*@UnknownInitialization(Foo.class) Foo Foo.this*/) {
                                                  ^
  required: <anonymous Runnable>
  found: Foo
```
Thanks!

**Comment by wmdietl** [comment]
This looks like Issue #266 
The current best work-around is to create a new final local variable after initializing the field. See https://github.com/typetools/checker-framework/issues/266#issuecomment-118318478 for an example.

**Comment by mihai-dumitresq** [comment]
I was about to mention #266 since it's similar but different in an important way. The access variable is a field rather than a local variable.
Which means that **even if #266 is fixed this case would still fail**.

I admit that the workaround is similar (final local variable).

**Comment by wmdietl** [comment]
The correct fix will propagate current flow-refinements into the inner scope of the anonymous class. It doesn't matter whether whether the refinement is for a local variable or a field.

Thanks for submitting these issues and please keep them coming!

**Comment by mihai-dumitresq** [comment]
Maybe this example will make the difference clearer:

```Java
import com.google.common.util.concurrent.SettableFuture;
class Foo {
  private final SettableFuture<Bar> mBar;
  private final Runnable mRunnable =
        new Runnable() {
          @Override
          public void run() {
            // error: [dereference.of.nullable] dereference of possibly-null reference
            if (mBar.isCancelled()) {
              // whatever
            }
          }
        };

  public Foo() {
    mBar = SettableFuture.create();
    mBar.addListener(mRunnable, MoreExecutors.directExecutor());
  }
}
```

**Comment by mihai-dumitresq** [comment]
To explain a bit further: what I'm trying to say is that the problem here is that there's no way to make constraints on `this` when run() is executed (like using @RequiresNonNull or @UnknownInitialization).

**Comment by wmdietl** [comment]
I can't reproduce the problem. I made the following compilable program:

```
import com.google.common.util.concurrent.SettableFuture;
import java.util.concurrent.Executor;

class Foo {
  private final SettableFuture<Object> mBar;
  private final Runnable mRunnable =
        new Runnable() {
          @Override
          public void run() {
            if (mBar.isCancelled()) {
              // whatever
            }
          }
        };

  public Foo(Executor exec) {
    mBar = SettableFuture.create();
    mBar.addListener(mRunnable, exec);
  }
}
```

And it doesn't give an error from the Nullness Checker.
`mBar` is implicitly non-null.

**Comment by wmdietl** [comment]
However, your overall observation is valid. You can only propagate state into the anonymous class, if it is true in all possible executions.
For final fields that is the case. Once you checked that it is non-null, that fact can't change anymore.
More care needs to be taken for nullable, non-final fields.

**Comment by mihai-dumitresq** [comment]
Huh. I would have thought the two would be pretty much equivalent. Is the no complaints when using an mRunnable a known bug? Or should I file one for the mRunnable.run() below?

I've extracted out everything into the compilable case below (tested with v2.1.1 this time):
```Java
import com.google.common.util.concurrent.SettableFuture;
import org.checkerframework.checker.nullness.qual.RequiresNonNull;
import java.util.concurrent.Executor;

class Foo {
  private final SettableFuture<Object> mBar;
  private final Runnable mRunnable =
        new Runnable() {
          @Override
          public void run() {
            // No problem, even though it NPEs.
            if (mBar.isCancelled()) {
              // whatever
            }
          }
        };

  public Foo(Executor exec) {
    mRunnable.run();

    mBar = SettableFuture.create();
    mBar.addListener(mRunnable, exec);
    mBar.addListener(
        new Runnable() {
          @Override
          // error: [contracts.precondition.override.invalid] @NonNull void run(@Initialized @NonNull <anonymous Foo$2> this) in <anonymous Foo$2> cannot override @NonNull void run(@Initialized @NonNull Runnable this) in java.lang.Runnable; the precondition '@NonNull' for 'this.mBar' in the subclass method must be weaker (or equally strong) compared to the superclass method
          // @RequiresNonNull("mBar")
          public void run(
             // error: the receiver type does not match the enclosing class type
             // Maybe there is a correct syntax but I don't know it.
             // @UnknownInitialization(Foo.class) Foo Foo.this
          ) {
            // error: [dereference.of.nullable] dereference of possibly-null reference
            if (mBar.isCancelled()) {
              // whatever
            }
          }
        },
        exec);
  }
}
```

**Comment by wmdietl** [comment]
Yes, that's a bug! When checking the anonymous class in mRunnable, we use the declared type of mBar, which is non-null.
However, the `mRunnable.run()` invocation is in the constructor, when the field is not set yet.
This needs to correctly consider initialization.

Can you file a new issue without the whole SettableFuture overhead? And add a `main` method that actually triggers an NPE.

You cannot strengthen a precondition, which is why the added `@RequiresNonNull` fails.

**Comment by mihai-dumitresq** [comment]
Done, filed #904.

Now, to settle this issue. Do you want to reopen it and, make it about `@RequiresNonNull(mBar)`:
- either there isn't something to stengthen in which case mBar should be considered non null in which case why the error?
- there is something to stengthen in which case `@RequiresNonNull(mBar)` should work

**Comment by wmdietl** [comment]
You can't soundly strengthen pre-conditions. The Runnable interface doesn't require anything to be non-null. If your implementation adds a `@RequiresNonNull` there would be no way to enforce it.
I think with #266 and #904 all issues from this example are covered.

**Comment by mihai-dumitresq** [comment]
I see what you mean now. I'll add the following to the #266 thread to be used as a test case:

```Java
class Foo {
  private final Object mBar;

  public Foo() {
    mBar = "test";
    Runnable runnable = new Runnable() {
        @Override
        public void run() {
          // error: [dereference.of.nullable] dereference of possibly-null reference mBar
          mBar.toString();
        }
      };
    runnable.run();
  }
}
```

**Comment by wmdietl** [comment]
But why would you want that warning in this example?

**Comment by mihai-dumitresq** [comment]
I wouldn't want it, just wanted to show where the error is. :)

Anyway, thanks a lot for this discussion and all the help!

**Comment by wmdietl** [comment]
Can you edit your example and change the `// error:` to `// unexpected error here` or some such?
Then it's clearer what you mean.

Thanks for finding these issues!

**Comment by mihai-dumitresq** [comment]
Done.
